nexus/types/src/internal_api/views.rs

```diff
@@ -42,7 +44,16 @@ impl From<steno::SagaView> for Saga {
 pub enum SagaState {
     Running,
     Succeeded,
-    Failed { error_node_name: steno::NodeName, error_info: SagaErrorInfo },
+    Failed {
+        error_node_name: steno::NodeName,
+        error_info: SagaErrorInfo,
+    },
+    Stuck {
```

Comment:
I'm not too clear on the distinction between these variants. I gather failed implies a forward action failed, and stuck means an undo failed. Is the intent to convey that a `Failed` saga could be retried, but a `Stuck` saga can't? I guess I'm wondering the degree to which the variant names convey what happened (action failed, undo action failed, etc) as opposed to what can be done for a saga now, given that it's in either of these states.

Does that make sense?

Reply:
First, just to be clear about scope, this is a debug-only API exposed only on the internal network for operators to inspect saga state.

I've been a little on the fence about how to describe sagas whose undo actions fail.  Normally, the invariant provided by sagas is that either you wind up at the end state (all the actions having completed successfully) or you wind up back at the initial state (all the undo actions having been applied).  In this case, we've gotten stuck somewhere in the middle.  Some of the actions have been applied and not undone.  It's pretty different from a "failed" saga (that was successfully unwound) because the system has come to rest broken.  I wanted to elevate that here to its own state.  In the limit, I imagine we'll have queries that report the Stuck sagas and we'll want to open support cases or something.

At first I took this same approach with steno#138, having "Stuck" be a new terminal state alongside "Done".  But I found that it complicated a lot of stuff without buying much.  So Steno _does_ expose these sagas as "'done', but with a slightly different error variant" rather than "come to rest 'stuck' rather than 'done'".  So there's a bit of a difference between Steno and Omicron's debug API here.  I decided to go with it and see how we actually wind up using this.